File: spark-df-ds/src/main/scala/spark/dataFrame/Schema.py
import findspark
findspark.init()
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql import Row

#parquet,avro,json,csv and orc file will have infer schema --> .option("header","true")

spark = SparkSession.builder.master("local").appName("anil").getOrCreate()
sc = spark.sparkContext
sc.setLogLevel("off")
class Schema :

  def structtype(self):

    schemaString = "name age"
    #fields = map(lambda fieldName:  StructField(fieldName, StringType(),True), schemaString.split(" "))
    fields =  [StructField(fieldName, StringType(),True) for fieldName in schemaString.split(" ")]
    return StructType(fields) #StructType(StructField(name,StringType,true), StructField(age,StringType,true))

  def hardCode(sels):

    lines = sc.textFile("C:\\Users\\me\\PycharmProjects\\PyThon\\src\\main\\resource\\people.txt")
    parts = lines.map(lambda l: l.split(","))
    schemaPeople = spark.createDataFrame(parts).toDF("name", "age")
    schemaPeople.show(truncate=False)

  def coll_creatdf (self):
     #coll = ["name", "age"]
     lines = sc.textFile("C:\\Users\\me\\PycharmProjects\\PyThon\\src\\main\\resource\\people.txt")
     parts = lines.map(lambda l: l.split(","))
     schemaPeople = spark.createDataFrame(parts,["name", "age"])
     schemaPeople.show(truncate=False)

  def coll_todf(sels):
    coll = ["name", "age"]
    lines = sc.textFile("C:\\Users\\me\\PycharmProjects\\PyThon\\src\\main\\resource\\people.txt")
    parts = lines.map(lambda l: l.split(","))
    schemaPeople = spark.createDataFrame(parts).toDF(*coll)
    #schemaPeople = spark.createDataFrame(parts).toDF(*["name", "age"])
    schemaPeople.show(truncate=False)

  def schems_creatdf (self):
     #coll = ["name", "age"]
     lines = sc.textFile("C:\\Users\\me\\PycharmProjects\\PyThon\\src\\main\\resource\\people.txt")
     parts = lines.map(lambda l: l.split(","))
     schemaPeople = spark.createDataFrame(parts, Schema().structtype())
     schemaPeople.show()

  def map (self):

    lines = sc.textFile("C:\\Users\\me\\PycharmProjects\\PyThon\\src\\main\\resource\\people.txt")
    parts = lines.map(lambda l: l.split(","))
    people = parts.map(lambda p: Row(name=p[0], age=int(p[1])))
    schemaPeople = spark.createDataFrame(people)
    schemaPeople.show(truncate=False)

  def schema(self):
   df = spark.read\
    .option("delimiter",",")\
    .schema(Schema().structtype().add("_corrupt_record",StringType(),True))\
    .csv("C:\\Users\\me\\PycharmProjects\\PyThon\\src\\main\\resource\\people.txt")
   df.show(truncate=False)


struct1 = StructType().add("f1", StringType(), True).add("f2", StringType(), True, None)

struct2 = StructType([StructField("f1", StringType(), True),
                      StructField("f2", StringType(), True, None)])

File: spark-df-ds/src/main/scala/spark/dataFrame/Schema.scala
package spark.dataFrame


import java.io.File

import org.apache.spark.sql._
import org.apache.spark.sql.types._

import scala.io.Source
//parquet,avro,json,csv and orc file will have infer schema --> .option("header","true")
object Schema extends Serializable {

  def structType: StructType ={

    val schemaString = "name age"
    val fields  = schemaString.split(" ")
      .map(fieldName => StructField(fieldName, StringType, nullable = true))
   StructType(fields) //StructType(StructField(name,StringType,true), StructField(age,StringType,true))

  }
  def seq: Seq[String] ={
      Seq("name", "age")
  }
  case class Person(name: String, age: Long)
  def structtype_from_json: StructType ={

    val schemaSource = Source.fromFile(new File("C:\\Users\\me\\IdeaProjects\\bigData\\src\\main\\resources\\schema.json")).getLines.mkString
    DataType.fromJson(schemaSource).asInstanceOf[StructType]

  }
  def array_and_map_type: Unit ={
  val arrayStructureData = Seq(
    Row(Row("James ","","Smith"),List("Cricket","Movies"),Map("hair"->"black","eye"->"brown")),
    Row(Row("Michael ","Rose",""),List("Tennis"),Map("hair"->"brown","eye"->"black")),
    Row(Row("Robert ","","Williams"),List("Cooking","Football"),Map("hair"->"red","eye"->"gray")),
    Row(Row("Maria ","Anne","Jones"),null,Map("hair"->"blond","eye"->"red")),
    Row(Row("Jen","Mary","Brown"),List("Blogging"),Map("white"->"black","eye"->"black"))
  )

  val arrayStructureSchema = new StructType()
    .add("name",new StructType()
      .add("firstname",StringType)
      .add("middlename",StringType)
      .add("lastname",StringType))
    .add("hobbies", ArrayType(StringType))
    .add("properties", MapType(StringType,StringType))

}
  /*def conver_case_class_into_structType: Unit ={

    case class Name(first:String,last:String,middle:String)
    case class Employee(fullName:Name,age:Integer,gender:String)

    import org.apache.spark.sql.catalyst.ScalaReflection
    val schema = ScalaReflection.schemaFor[Employee].dataType.asInstanceOf[StructType]

    val encoderSchema = Encoders.product[Employee].schema
    encoderSchema.printTreeString()
  }*/
  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder().appName("read_and_write_hdfs").master("local").getOrCreate()
    val structureData = Seq(
      Row(Row("James ","","Smith"),"36636","M",3100),
      Row(Row("Michael ","Rose",""),"40288","M",4300),
      Row(Row("Robert ","","Williams"),"42114","M",1400),
      Row(Row("Maria ","Anne","Jones"),"39192","F",5500),
      Row(Row("Jen","Mary","Brown"),"","F",-1)
    )
    val df3 = spark.createDataFrame(spark.sparkContext.parallelize(structureData),structtype_from_json)
    df3.repartition(1)
      .write
      .format("json")
      .mode(SaveMode.Overwrite)
      .save("C:\\Users\\me\\IdeaProjects\\bigData\\src\\main\\resources\\Array.json")
  }

}
